lex.go

```diff
@@ -226,3 +319,147 @@ func tokenize(reader io.Reader, tokenCh chan NgxToken) {
+		"balancer_by_lua_block",
+		"ssl_client_hello_by_lua_block",
+		"ssl_certificate_by_lua_block",
+		"ssl_session_fetch_by_lua_block",
+		"ssl_session_store_by_lua_block",
+	}
+}
+
+//nolint:funlen,gocognit,gocyclo,nosec
+func (ll *LuaLexer) Lex(matchedToken string) <-chan NgxToken {
+	tokenCh := make(chan NgxToken)
+
+	tokenDepth := 0
+
+	go func() {
+		defer close(tokenCh)
+		var tok strings.Builder
+		var inQuotes bool
+
+		// special handling for'set_by_lua_block' directive
+		// ignore potential hardcoded credentials linter warning for "set_by_lua_block"
+		if matchedToken == "set_by_lua_block" /* #nosec G101 */ {
+			arg := ""
+			for {
+				if !ll.s.Scan() {
+					return
+				}
+				next := ll.s.Text()
+				if isSpace(next) {
+					if arg != "" {
+						tokenCh <- NgxToken{Value: arg, Line: ll.s.Line(), IsQuoted: false}
+						break
+					}
+
+					for isSpace(next) {
+						if !ll.s.Scan() {
+							return
+						}
+						next = ll.s.Text()
+					}
+				}
+				arg += next
+			}
+		}
+
+		// check that Lua block starts correctly
+		for {
+			if !ll.s.Scan() {
+				return
+			}
+			next := ll.s.Text()
+
+			if !isSpace(next) {
+				if next != "{" {
+					lineno := ll.s.Line()
+					tokenCh <- NgxToken{Error: &ParseError{File: &lexerFile, What: `expected "{" to start lua block`, Line: &lineno}}
+					return
+				}
+				tokenDepth++
+				break
+			}
+		}
+
+		// Grab everything in Lua block as a single token and watch for curly brace '{' in strings
+		for {
+			if !ll.s.Scan() {
+				return
+			}
+
+			next := ll.s.Text()
+			if err := ll.s.Err(); err != nil {
+				lineno := ll.s.Line()
+				tokenCh <- NgxToken{Error: &ParseError{File: &lexerFile, What: err.Error(), Line: &lineno}}
+			}
+
+			switch {
+			case next == "{" && !inQuotes:
+				tokenDepth++
+				if tokenDepth > 1 { // not the first open brace
+					tok.WriteString(next)
+				}
+
+			case next == "}" && !inQuotes:
+				tokenDepth--
+				if tokenDepth < 0 {
+					lineno := ll.s.Line()
+					tokenCh <- NgxToken{Error: &ParseError{File: &lexerFile, What: `unexpected "}"`, Line: &lineno}}
+					return
+				}
+
+				if tokenDepth > 0 { // not the last close brace for it to be 0
+					tok.WriteString(next)
+				}
+
+				if tokenDepth == 0 {
+					tokenCh <- NgxToken{Value: tok.String(), Line: ll.s.Line(), IsQuoted: true}
+					tokenCh <- NgxToken{Value: ";", Line: ll.s.Line(), IsQuoted: false} // For an end to the Lua string based on the nginx bahavior
+					// See: https://github.com/nginxinc/crossplane/blob/master/crossplane/ext/lua.py#L122C25-L122C41
+					return
+				}
+
+			case next == `"` || next == "'":
+				inQuotes = !inQuotes
+				tok.WriteString(next)
+
+			default:
+				// Expected first token encountered to be a "{" to open a Lua block. Handle any other non-whitespace
+				// character to mean we are not about to tokenize Lua
+				// ignoring everything until first open brace where tokenDepth > 0
+				if isSpace(next) && tokenDepth == 0 {
+					continue
+				}
```

Comment:
Can you explain what is meant by "we are not about to tokenize Lua"? I'm not sure what this is for exactly.

Reply:
Oh, ok! Maybe if the comment called out the case if it's that specific it would be clearer?

Reply:
done